readme.md
# Events Node.js SDK

This SDK lets you track user event data from your Node applications.

## Installing the SDK

Run the following command to install the Node.js SDK via **npm**:

```bash
$ npm install @ht-sdks/events-sdk-node
```

## Using the SDK

Create a global client:

```javascript
const Analytics = require('@ht-sdks/events-sdk-node');

const client = new Analytics(WRITE_KEY, {
  dataPlaneUrl: DATA_PLANE_URL,
});
```

## SDK Initialization Options

Below parameters are optional and can be passed during SDK initialization:

| Name                   | Type     | Default value                            | Description                                                                                       |
| :--------------------- | :------- | :--------------------------------------- | :------------------------------------------------------------------------------------------------ |
| `dataPlaneUrl`         | String   |                                          | The data plane URL.                                                                               |
| `path`                 | String   | `/v1/batch`                              | Path to batch endpoint.                                                                           |
| `flushAt`              | Number   | 20                                       | The number of events to be flushed when reached this limit.                                       |
| `flushInterval`        | Number   | 10000                                    | The maximum timespan (in milliseconds) after which the events from the in-memory queue is flushed |
| `maxQueueSize`         | Number   | 460800(500kb)                            | Maximum payload size of a batch request                                                           |
| `maxInternalQueueSize` | Number   | 20000                                    | The maximum length of the in-memory queue                                                         |
| `logLevel`             | String   | 'info'                                   | Log level. `Ex: 'debug', 'error'`                                                                 |
| `axiosConfig`          | Object   | N/A                                      | Axios config                                                                                      |
| `axiosInstance`        | Object   | N/A                                      | Axios instance                                                                                    |
| `axiosRetryConfig`     | Object   | N/A                                      | Axios retry configuration                                                                         |
| `retryCount`           | Number   | 3                                        | Number of times requests will be retried by axios if failed                                       |
| `errorHandler`         | Function | N/A                                      | A function that will be called if request to server failed                                        |
| `gzip`                 | Boolean  | true                                     | Whether to compress request with gzip or not                                                      |

## Example calls

> Unlike the client-side SDKs that deal with a single user at a given time, the server-side SDKs deal with multiple users simultaneously. Therefore, you must specify either the userId or anonymousId every time while making any API calls supported by the Node SDK.

Identify
```javascript
client.identify({
  userId: "1hKOmRA4GRlm",
  traits: {
    name: "Alex Keener",
    email: "alex@example.com",
    plan: "Free",
    friends: 21,
  },
})
```

Track
```javascript
client.track({
  userId: "1hKOmRA4GRlm",
  event: "Item Viewed",
  properties: {
    revenue: 19.95,
    shippingMethod: "Premium",
  },
})
```

Page
```javascript
client.page({
  userId: "1hKOmRA4GRlm",
  category: "Food",
  name: "Pizza",
  properties: {
    url: "https://example.com",
    title: "Pizza",
    referrer: "https://google.com",
  },
})
```

Screen
```javascript
client.screen({
  userId: "12345",
  category: "Food",
  name: "Pizza",
  properties: {
    screenSize: 10,
    title: "Pizza",
    referrer: "https://google.com",
  },
})
```

Group
```javascript
client.group({
  userId: "12345",
  groupId: "1",
  traits: {
    name: "Company",
    description: "Google",
  },
})
```

Alias
```javascript
client.alias({
  previousId: "old_id",
  userId: "new_id",
})
```


CHANGELOG.md
# Changelog

### [1.0.2](https://github.com/ht-sdks/events-sdk-node/compare/v1.0.1...v1.0.2) (2023-09-19)
* update default API endpoints

### [1.0.1](https://github.com/ht-sdks/events-sdk-node/compare/v1.0.0...v1.0.1) (2023-08-31)
* testing release flow

### [1.0.0](https://github.com/ht-sdks/events-sdk-node/compare/v0.0.0...v1.0.0) (2023-08-31)
* updated default data plane and control plane URLs
* updated documentation
* updated naming


---
Pre Fork:
---

### [2.0.4](https://github.com/rudderlabs/rudder-sdk-node/compare/v2.0.3...v2.0.4) (2023-08-28)


### Bug Fixes

* concurrent event triggering unnecessary flush call ([#113](https://github.com/rudderlabs/rudder-sdk-node/issues/113)) ([5dd8999](https://github.com/rudderlabs/rudder-sdk-node/commit/5dd89999a7625c999434016dbf3bd6d509d7417a))
* remove winston logger dependency to enable serverless environments ([#112](https://github.com/rudderlabs/rudder-sdk-node/issues/112)) ([3619247](https://github.com/rudderlabs/rudder-sdk-node/commit/36192471055fdfc2032223986267715c2b48d4c4))

### [2.0.3](https://github.com/rudderlabs/rudder-sdk-node/compare/v2.0.2...v2.0.3) (2023-06-30)


### Bug Fixes

* errorhandler type declaration ([#101](https://github.com/rudderlabs/rudder-sdk-node/issues/101)) ([e7ee335](https://github.com/rudderlabs/rudder-sdk-node/commit/e7ee335015fba54de6c7a1c29c2dd7438c79a566))

## [v2.0.2](https://github.com/rudderlabs/rudder-sdk-node/compare/v2.0.1...v2.0.2)

### Fixes

- flush after a single event is sent [`#93`](https://github.com/rudderlabs/rudder-sdk-node/pull/93)
- added more loggers to assist debugging [`#94`](https://github.com/rudderlabs/rudder-sdk-node/pull/94)

## [v2.0.1](https://github.com/rudderlabs/rudder-sdk-node/compare/v2.0.0...v2.0.1)

### Fixes

- library info override when provided in context [`#89`](https://github.com/rudderlabs/rudder-sdk-node/pull/89)

## [v2.0.0](https://github.com/rudderlabs/rudder-sdk-node/compare/v1.1.4...v2.0.0)

### Features

- Allow dataPlaneUrl to be be passed in options instead as an argument in initialization. [`#81`](https://github.com/rudderlabs/rudder-sdk-node/pull/81)
- Allow passing axios instance or config in options. [`#81`](https://github.com/rudderlabs/rudder-sdk-node/pull/81)
- Allow passing error handler function in options. [`#81`](https://github.com/rudderlabs/rudder-sdk-node/pull/81)
- Allow to configure batch events path from initialization [`#81`](https://github.com/rudderlabs/rudder-sdk-node/pull/81)
- added request compression feature with gzip [`#81`](https://github.com/rudderlabs/rudder-sdk-node/pull/81)
- made bull package as an optional dependency [`#81`](https://github.com/rudderlabs/rudder-sdk-node/pull/81)
- added type declaration for screen api [`#81`](https://github.com/rudderlabs/rudder-sdk-node/pull/81)

### Fixes

- default queue prefix [`#81`](https://github.com/rudderlabs/rudder-sdk-node/pull/81)
- Cli app to use latest version of dependency [`#81`](https://github.com/rudderlabs/rudder-sdk-node/pull/81)

### BREAKING CHANGES

- Allow dataPlaneUrl to be passed in options instead as an argument in initialization and must no longer include '/v1/batch' part. [`#81`](https://github.com/rudderlabs/rudder-sdk-node/pull/81)

## [v1.1.4](https://github.com/rudderlabs/rudder-sdk-node/compare/v1.1.3...v1.1.4)

### Merged

- fix: type declaration for createPersistenceQueue API [`#72`](https://github.com/rudderlabs/rudder-sdk-node/pull/72)

## [v1.1.3](https://github.com/rudderlabs/rudder-sdk-node/compare/v1.1.2...v1.1.3) - 2022-05-06

### Merged

- patch: add default 'channel' field to event [`#70`](https://github.com/rudderlabs/rudder-sdk-node/pull/70)

### Commits

- Regenerated History.md [`190d07a`](https://github.com/rudderlabs/rudder-sdk-node/commit/190d07aa3306e0ace33eb864bbb2b662552db876)

## [v1.1.2](https://github.com/rudderlabs/rudder-sdk-node/compare/v1.1.1...v1.1.2) - 2022-04-05

### Merged

- TS declarations update [`#68`](https://github.com/rudderlabs/rudder-sdk-node/pull/68)
- Cleaned up package.json [`#66`](https://github.com/rudderlabs/rudder-sdk-node/pull/66)

## [v1.1.1](https://github.com/rudderlabs/rudder-sdk-node/compare/v1.1.0...v1.1.1) - 2022-03-03

### Merged

- Bump follow-redirects from 1.14.7 to 1.14.8 [`#63`](https://github.com/rudderlabs/rudder-sdk-node/pull/63)
- Bump ajv from 6.10.0 to 6.12.6 [`#62`](https://github.com/rudderlabs/rudder-sdk-node/pull/62)

### Commits

- package-lock.json not ignored anymore. Removed yarn.lock [`ed2bf14`](https://github.com/rudderlabs/rudder-sdk-node/commit/ed2bf14e8a2c8c31ab32a6a3c2e1b620cf119e4e)
- Upgraded all the packages to latest and cleaned-up package.json [`11a0789`](https://github.com/rudderlabs/rudder-sdk-node/commit/11a0789bfbf589a37cd484a698902e563bf96a36)
- Regenerated History.md [`14f2d85`](https://github.com/rudderlabs/rudder-sdk-node/commit/14f2d855d66b3bcae1e9eae62eab298862574592)

## [v1.1.0](https://github.com/rudderlabs/rudder-sdk-node/compare/v1.0.14...v1.1.0) - 2022-02-24

### Merged

- concurrency added for flush [`#64`](https://github.com/rudderlabs/rudder-sdk-node/pull/64)

### Commits

- Update readme.md [`4ec186f`](https://github.com/rudderlabs/rudder-sdk-node/commit/4ec186fbb84ae415c1c24213ab7bbbb30d0e54fc)
- Updated History.md to latest [`9a825cd`](https://github.com/rudderlabs/rudder-sdk-node/commit/9a825cd41a721bd8a4d3e932967a2298614bb91f)

## [v1.0.14](https://github.com/rudderlabs/rudder-sdk-node/compare/v1.0.13...v1.0.14) - 2022-01-27

### Merged

- Bump node-fetch from 2.6.1 to 2.6.7 [`#59`](https://github.com/rudderlabs/rudder-sdk-node/pull/59)
- Auto-changelog package updated [`#58`](https://github.com/rudderlabs/rudder-sdk-node/pull/58)
- Bump trim-off-newlines from 1.0.1 to 1.0.3 [`#55`](https://github.com/rudderlabs/rudder-sdk-node/pull/55)
- Bump follow-redirects from 1.14.5 to 1.14.7 [`#51`](https://github.com/rudderlabs/rudder-sdk-node/pull/51)

### Commits


src/Logger.js
const LOG_MSG_PREFIX = 'Hightouch';
const LOG_LEVEL_MAP = {
  log: 0,
  info: 1,
  debug: 2,
  warn: 3,
  error: 4,
  none: 5,
};

/**
 * Service to log messages/data to output provider, default is console
 */
class Logger {
  constructor(minLogLevel = 4) {
    this.minLogLevel = minLogLevel;
    this.logProvider = console;
  }

  log(...data) {
    this.outputLog('log', data);
  }

  info(...data) {
    this.outputLog('info', data);
  }

  debug(...data) {
    this.outputLog('debug', data);
  }

  warn(...data) {
    this.outputLog('warn', data);
  }

  error(...data) {
    this.outputLog('error', data);
  }

  outputLog(logMethod, data) {
    if (this.minLogLevel <= LOG_LEVEL_MAP[logMethod]) {
      this.logProvider[logMethod.toLowerCase()](...this.formatLogData(logMethod, data));
    }
  }

  /**
   * Formats the console message
   */
  // eslint-disable-next-line class-methods-use-this
  formatLogData(level, data) {
    if (Array.isArray(data) && data.length > 0) {
      // trim whitespaces for original message
      const originalMsg = typeof data[0] === 'string' ? data[0].trim() : '';

      // prepare the final message
      const timestamp = new Date().toISOString();
      const msg = `${timestamp} [${LOG_MSG_PREFIX}] ${level}: ${originalMsg}`;

      const styledLogArgs = [msg];

      // add first it if it was not a string msg
      if (typeof data[0] !== 'string') {
        styledLogArgs.push(data[0]);
      }

      // append rest of the original arguments
      styledLogArgs.push(...data.slice(1));
      return styledLogArgs;
    }

    return data;
  }
}

module.exports = {
  Logger,
  LOG_LEVEL_MAP,
};


examples/sample-app/app.js
const HTEvents = require('@ht-sdks/events-sdk-node');
require('dotenv').config({ path: '../../.env' });

const writeKey = process.env.WRITE_KEY;
const dataPlaneUrl = process.env.DATAPLANE_URL;

const client = new HTEvents(writeKey, {
  dataPlaneUrl,
  flushAt: 2,
  logLevel: 'debug',
});
/**
 * Sample function to send 3 events[identify,track,track] and make sure events are flowing
 */
function test() {
  client.identify(
    {
      userId: 'Test user 1',
      traits: {
        name: 'Name Username',
        email: 'name@website.com',
        plan: 'Free',
        friends: 21,
      },
    },
    () => {
      console.log('In identify call');
    },
  );

  client.track(
    {
      userId: 'Test user 1',
      event: 'Item Purchased',
      properties: {
        revenue: 19.95,
        shippingMethod: 'Premium',
      },
    },
    () => {
      console.log('In track call');
    },
  );
  client.track(
    {
      userId: 'Test user 2',
      event: 'Item Viewed',
      properties: {
        price: 45,
        currency: 'USD',
        productId: 'Product-12345',
      },
    },
    () => {
      console.log('In track call 2');
    },
  );

  // await client.flush();
}

try {
  test();
} catch (e) {
  console.log(e.message);
}

exports.HTEvents = HTEvents;

// run this file with the command "node app"


index.d.ts
export default Analytics;

/**
 * Represents a generic object in the APIs
 * Use for parameters like context, traits etc.
 */
export interface apiObject {
  [index: string]:
    | string
    | number
    | boolean
    | undefined
    | apiObject
    | (string | number | boolean | apiObject)[];
}

/**
 * Represents the integration options object
 * Example usages:
 * integrationOptions { All: false, "Google Analytics": true, "Braze": true}
 * integrationOptions { All: true, "Chartbeat": false, "Customer.io": false}
 */
export interface integrationOptions {
  // Defaults to true
  // If set to false, specific integration should be set to true to send the event
  All?: boolean;
  // Destination name: true/false/integration specific information
  [index: string]: boolean | apiObject | undefined;
}

/**
 * Represents the constructor options object
 * Example usages:
 * constructorOptions { flushAt: 20, "flushInterval": 20000, "enable": true, "maxQueueSize":20000, "logLevel": "info"/"debug"/"error"/"silly"}
 */
export interface constructorOptions {
  flushAt?: number;
  flushInterval?: number;
  enable?: boolean;
  maxQueueSize?: number;
  maxInternalQueueSize?: number;
  logLevel?: string;
  dataPlaneUrl?: string;
  host?: string;
  path?: string;
  axiosConfig?: any;
  axiosInstance?: any;
  axiosRetryConfig?: any;
  retryCount?: number;
  errorHandler?: (error?: any) => void;
  gzip?: boolean;
}

/**
 * Represents the callback in the APIs
 */
export type apiCallback = () => void;
declare class Analytics {
  /**
   * Initialize a new `Analytics` with your project's `writeKey` and an
   * optional dictionary of `options`.
   *
   * @param {String} writeKey
   * @param {Object} [options] (optional)
   *   @property {Number} [flushAt] (default: 20)
   *   @property {Number} [flushInterval] (default: 10000)
   *   @property {Number} [maxQueueSize] (default: 500 kb)
   *   @property {Number} [maxInternalQueueSize] (default: 20000)
   *   @property {String} [logLevel] (default: 'info')
   *   @property {String} [dataPlaneUrl] (default: 'https://events.us-east-1.hightouch.com')
   *   @property {String} [host] (default: 'https://events.us-east-1.hightouch.com')
   *   @property {String} [path] (default: '/v1/batch')
   *   @property {Boolean} [enable] (default: true)
   *   @property {Object} [axiosConfig] (optional)
   *   @property {Object} [axiosInstance] (default: axios.create(options.axiosConfig))
   *   @property {Object} [axiosRetryConfig] (optional)
   *   @property {Number} [retryCount] (default: 3)
   *   @property {Function} [errorHandler] (optional)
   *   @property {Boolean} [gzip] (default: true)
   */
  constructor(writeKey: string, options?: constructorOptions);

  /**
   *
   * @param {Object} queueOpts
   * @param {String=} queueOpts.queueName
   * @param {String=} queueOpts.prefix
   * @param {Boolean=} queueOpts.isMultiProcessor
   * @param {Object} queueOpts.redisOpts
   * @param {Number=} queueOpts.redisOpts.port
   * @param {String=} queueOpts.redisOpts.host
   * @param {Number=} queueOpts.redisOpts.db
   * @param {String=} queueOpts.redisOpts.password
   * @param {Object=} queueOpts.jobOpts
   * @param {Number} queueOpts.jobOpts.maxAttempts
   * {
   *    queueName: string = hightouchEventsQueue,
   *    prefix: string = hightouch
   *    isMultiProcessor: booloean = false
   *    redisOpts: {
   *      port?: number = 6379;
   *      host?: string = localhost;
   *      db?: number = 0;
   *      password?: string;
   *    },
   *    jobOpts: {
   *      maxAttempts: number = 10
   *    }
   * }
   * @param {*} callback
   *  All error paths from redis and queue will give exception, so they are non-retryable from SDK perspective
   *  The queue may not function for unhandled promise rejections
   *  this error callback is called when the SDK wants the user to retry
   */
  createPersistenceQueue(
    queueOpts: {
      queueName?: string;
      prefix?: string;
      isMultiProcessor?: boolean;
      redisOpts: {
        port?: number;
        host?: string;
        db?: number;
        password?: string;
      };
      jobOpts?: {
        maxAttempts?: number;
      };
    },
    callback: (error?: Error | string) => void,
  ): void;

  /**
   * Send an identify `message`.
   *
   * @param {Object} message
   * @param {String=} message.userId (optional)
   * @param {String=} message.anonymousId (optional)
   * @param {Object=} message.context (optional)
   * @param {Object=} message.traits (optional)
   * @param {Object=} message.integrations (optional)
   * @param {Date=} message.timestamp (optional)
   * @param {Function=} callback (optional)
   * @return {Analytics}
   */
  identify(
    message: {
      userId?: string;
      anonymousId?: string;
      context?: apiObject;
      traits?: apiObject;
      integrations?: integrationOptions;
      timestamp?: Date;
    },
    callback?: apiCallback,
  ): Analytics;
  /**
   * Send a group `message`.
   *
   * @param {Object} message
   * @param {String} message.groupId
   * @param {String=} message.userId (optional)
   * @param {String=} message.anonymousId (optional)
   * @param {Object=} message.context (optional)
   * @param {Object=} message.traits (optional)
   * @param {Object=} message.integrations (optional)
   * @param {Date=} message.timestamp (optional)
   * @param {Function=} callback (optional)
   * @return {Analytics}
   */
  group(
    message: {
      groupId: string;
      userId?: string;
      anonymousId?: string;
      context?: apiObject;
      traits?: apiObject;
      integrations?: integrationOptions;
      timestamp?: Date;
    },
    callback?: apiCallback,
  ): Analytics;
  /**
   * Send a track `message`.
   *
   * @param {Object} message
   * @param {String} message.event
   * @param {String=} message.userId (optional)
   * @param {String=} message.anonymousId (optional)
   * @param {Object=} message.context (optional)
   * @param {Object=} message.properties (optional)
   * @param {Object=} message.integrations (optional)
   * @param {Date=} message.timestamp (optional)
   * @param {Function=} callback (optional)
   * @return {Analytics}
   */
  track(
    message: {
      event: string;
      userId?: string;
      anonymousId?: string;
      context?: apiObject;
      properties?: apiObject;
      integrations?: integrationOptions;
      timestamp?: Date;
    },
    callback?: apiCallback,
  ): Analytics;

  /**
   * Send a page `message`.
   *
   * @param {Object} message
   * @param {String} message.name
   * @param {String=} message.userId (optional)
   * @param {String=} message.anonymousId (optional)
   * @param {Object=} message.context (optional)
   * @param {Object=} message.properties (optional)
   * @param {Object=} message.integrations (optional)
   * @param {Date=} message.timestamp (optional)
   * @param {Function=} callback (optional)
   * @return {Analytics}
   */
  page(
    message: {
      name: string;
      userId?: string;
      anonymousId?: string;
      context?: apiObject;
      properties?: apiObject;
      integrations?: integrationOptions;
      timestamp?: Date;
    },
    callback?: apiCallback,
  ): Analytics;

  /**
   * Send a screen `message`.
   *
   * @param {Object} message
   * @param {String} message.name
   * @param {String=} message.userId (optional)
   * @param {String=} message.anonymousId (optional)
   * @param {Object=} message.context (optional)
   * @param {Object=} message.properties (optional)
   * @param {Object=} message.integrations (optional)
   * @param {Date=} message.timestamp (optional)
   * @param {Function=} callback (optional)
   * @return {Analytics}
   */
  screen(
    message: {
      name: string;
      userId?: string;
      anonymousId?: string;
      context?: apiObject;
      properties?: apiObject;
      integrations?: integrationOptions;
      timestamp?: Date;
    },
    callback?: apiCallback,
  ): Analytics;

  /**
   * Send an alias `message`.
   *
   * @param {Object} message
   * @param {String} message.previousId
   * @param {String=} message.userId (optional)
   * @param {String=} message.anonymousId (optional)
   * @param {Object=} message.context (optional)
   * @param {Object=} message.properties (optional)
   * @param {Object=} message.integrations (optional)
   * @param {Date=} message.timestamp (optional)
   * @param {Function=} callback (optional)
   * @return {Analytics}
   */
  alias(
    message: {
      previousId: string;
      userId?: string;
      anonymousId?: string;
      context?: apiObject;
      properties?: apiObject;
      integrations?: integrationOptions;
      timestamp?: Date;
    },
    callback?: apiCallback,
  ): Analytics;

  /**
   * Flush the current queue
   *
   * @param {Function} [callback] (optional)
   * @return {Analytics}
   */
  flush(callback?: Function): Analytics;
}
